server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java

```diff
@@ -216,12 +217,14 @@ public final int hashCode() {
      * @return the same input object or a {@link BytesRef} representation if input was of type string
      */
     static Object maybeConvertToBytesRef(Object obj) {
-        if (obj instanceof String) {
-            return BytesRefs.checkIndexableLength(BytesRefs.toBytesRef(obj));
-        } else if (obj instanceof CharBuffer) {
-            return BytesRefs.checkIndexableLength(new BytesRef((CharBuffer) obj));
-        } else if (obj instanceof BigInteger) {
-            return BytesRefs.toBytesRef(obj);
+        if (obj instanceof String v) {
+            byte[] b = new byte[UnicodeUtil.calcUTF16toUTF8Length(v, 0, v.length())];
+            UnicodeUtil.UTF16toUTF8(v, 0, v.length(), b);
+            return BytesRefs.checkIndexableLength(new BytesRef(b, 0, b.length));
+        } else if (obj instanceof CharBuffer v) {
+            return BytesRefs.checkIndexableLength(new BytesRef(v));
+        } else if (obj instanceof BigInteger v) {
+            return BytesRefs.toBytesRef(v);
```

Comment:
can we test the change?

Reply:
I though we could not because of Java, but actually we can and I implemented it [here](https://github.com/elastic/elasticsearch/pull/115655/commits/bd5c3211c4cde6f7e78d1f5003b09ab0a5e9a794).
Before my change the BytesRef could have length != bytes.length now it's the same and consistent. and always <= than previous length that was String#length*3